RIOT-OS/RIOT
# sys/atomic_utils: change API to also return previous value

### Contribution description

The read-modify-write operations exposed by `atomic_fetch_<OP>_u<WIDTH>()` functions currently in master don't return the previous value. However, this comes very much for free and makes this API much more versatile. The C11 atomics also do this. In fact, this is just human error that they don't do so in master.

This PR fixes the issue by adding the missing return value. This obviously is an API change, but users of this API usually don't need to be touched: They will just ignore the now added return value.

### Testing procedure

1. The unit test in `tests/sys_atomic_utils_unittests` have been extended to also cover the return value. Murdock will run those for you.
2. ~~There should be pretty much no difference in ROM size for `tests/sys_atomic_utils` and `tests/bench_sys_atomic_utils`~~
    - There are indeed no changes with `clang`, as it does optimize accesses to local (possibly register allocated) variables across `__asm__ volatile("":::"memory")` barriers. Those barriers are (rightfully!) placed in `irq_disable()` and `irq_restore()`. The reasoning is that the memory barrier does not apply to registers, as registers are not subject of concurrent access.
    - GCC is more strict here. Hence, the return value is always prepared even if it is unused, resulting in additional ROM.
    - To some degree the ROM increase can be reduced by hand crafted inline assembly implementations. But let's discuss whether the performance and ROM benefit is indeed worth having more inline ASM outside of this PR
3. ~~There shouldn't be a change in performance in the benchmark in `tests/bench_sys_atomic_utils`. (But beware: The error margin of that test is surprisingly high.)~~
    - For the same reasons as in 2., code slows down for GCC :-/

### Justification

Despite this sadly decreasing efficiency a bit for GCC (in terms of ROM and CPU cycles), this really should go in: Most of the use cases of the API rely on having the return value, especially the use case I added this API for. It was just human error that I forgot this in the first place.

Hence, the API might be leaner as it is in master now, but it is mostly useless. We really should have the return value, even at the expense of some overhead.

### Issues/PRs references

None
